test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -49,43 +55,62 @@ public static void main(String[] args) throws IOException {
             dc.configureBlocking(false);
             dc.register(sel, SelectionKey.OP_READ);
 
-            /* Test 1: not connected so ICMP port unreachable should not be received */
-            sendDatagram(dc, refuser);
-            int n = sel.select(2000);
-            if (n > 0) {
-                sel.selectedKeys().clear();
-                // BindException will be thrown if another service is using
-                // our expected refuser port, cannot run just exit.
-                DatagramChannel.open().bind(refuser).close();
-                throw new RuntimeException("Unexpected wakeup");
-            }
+            for (;;) {
+                /* Test 1: not connected so ICMP port unreachable should not be received */
+                sendDatagram(dc, refuser);
+                int n = sel.select(2000);
+                if (n > 0) {
+                    ignoreStrayWakeup = checkUnexpectedWakeup(dc, sel.selectedKeys());
+                    sel.selectedKeys().clear();
+                    if (ignoreStrayWakeup) {
```

Comment:
scenario 1 and 3 treat the received message differently to that of scenario 2 -- should they be handled the same ?

Reply:
I don't think so - scenario 2 is the one that expect the PortUnreachableException - so the wakeup is expected there and the read is expected to throw PortUnreadableException. The other two scenario are expected to timeout of the select with n == 0;

Reply:
I mean the actual testing of the received message -- scenario performs the read, and expects an Exception but if it gets a message,  which like the other scenarios it shouldn't, then the test of that message should be the same across all 3  scenarios ?
scenario 1 and 3 test if they receive what has been sent then retry, while scenario 2 will throw a RuntimeException if it receives what it has sent. This seems inconsistent.

Maybe change the send message to be completely unique to this test, there maybe other DatagramSocket or Channel tests that have message containing "Greetings"

Reply:
No I believe scenario 2 is different - but you are right - it would be more correct to replace lines 90-91 with this:

```
                    // BindException will be thrown if another service is using
                    // our expected refuser port, cannot run just exit.
                    DatagramChannel.open().bind(refuser).close();
                    throw new RuntimeException("PortUnreachableException not raised");
```
 
 In other words: we are connected - so we could only receive a message from something bound to the refuser's port. So check whether that port is still unbound, and if so, complain that PortUnreachableException should have been received.
Note that there still might be a race condition here because something could have transiently bound to the refuser and then released it. But the window of opportunity for the race should be smaller. 
Whether we should  `throw new RuntimeException("PortUnreachableException not raised");` or simply `continue` and try again could be discussed.

Reply:
> Maybe change the send message to be completely unique to this test, there maybe other DatagramSocket or Channel tests that have message containing "Greetings"

Good point. Although it's the only one I could find that uses Datagrams. We could add a time stamp - or a UUID for extra security.

Reply:
I think even something as simple as changing the message to ``Greetings from SelectWhenRefused!`` makes it unique, though I could do a UUID too?

Reply:
Good point too :-)